dali/kernels/signal/window/extract_windows_gpu.cuh

```diff
@@ -0,0 +1,568 @@
+ * @param dst           destination buffer
+ * @param num_windows   maximum number of windows to extract from given input
+ * @param stride        stride, in elements, between consecutive windows in `dst`
+ * @param src           input signal
+ * @param length        length, in samples, of the signal
+ * @param window        window function
+ * @param win_len       length, in samples, of the window
+ * @param win_center    center of the window function; typically win_len/2
+ * @param step          step, in samples, between consecutive windows in the input
+ * @param reflect       if true, reflect the signal at ends, otherwise zero-pad
+ */
+template <typename Dst, typename Src>
+__device__ void ExtractHorizontalWindows(
+    ptrdiff_t idx,
+    Dst *__restrict__ dst, ptrdiff_t num_windows, ptrdiff_t stride,
+    const Src *__restrict__ src, ptrdiff_t length,
+    const float *__restrict__ window,
+    int win_len,
+    int win_center,
+    int step,
+    bool reflect) {
+
+  // calculate the index of the first window that sample at `idx ` contributes to
+  ptrdiff_t idx0 = idx + win_center - win_len + step;  // add step to round up
+  ptrdiff_t win0 = ::max(ptrdiff_t(), idx0/step);
+
+  float value = 0;
+
+  if (reflect) {
+    ptrdiff_t src_idx = boundary::idx_reflect_101(idx, length);
+    value = ConvertSatNorm<float>(src[src_idx]);
+  } else {
+    if (idx >= 0 && idx < length)
+      value = ConvertSatNorm<float>(src[idx]);
+  }
+
+  for (int win_ofs = idx - (win0 * step - win_center), win_idx = win0;
+      win_ofs >= 0 && win_idx < num_windows;
+      win_ofs -= step, win_idx++) {
+    float v = window ? value * window[win_ofs] : value;
+    dst[win_idx * stride + win_ofs] = ConvertSatNorm<Dst>(v);
+  }
+}
+
+template <typename Dst, typename Src>
+__global__ void ExtractVerticalWindowsKernel(
+    Dst *__restrict__ dst, ptrdiff_t num_windows, ptrdiff_t stride,
+    const Src *__restrict__ src, ptrdiff_t length,
+    const float *__restrict__ window,
+    int win_len,
+    int win_center,
+    int step,
+    bool reflect) {
+  // This kernel reads kBlock elements from kBlock windows to shared memory
+  // and stores the transposed result to output in columns.
+
+  ExtractVerticalWindowsBlock(
+    blockIdx.x * kBlock,        // first window index
+    dst, num_windows, stride,   // output
+    src, length,                // input
+    window, win_len, win_center, step, reflect);  // windowing options
+}
+
+struct SampleDesc {
+  /// @brief Pointer to the beginning of the output window buffer
+  void *__restrict__ output;
+  /// @brief Total number of windows to extract from this signal
+  int num_windows;
+  /// @brief Stride between samples (for vertical windows) or between windows (for horizontal ones)
+  ptrdiff_t output_stride;
+  /// @brief Pointer to the beginning of the signal
+  const void *__restrict__ input;
+  /// @brief Length, in samples, of the input signal
+  ptrdiff_t length;
+};
+
+struct BlockDesc {
+  int sample_idx;
```

Comment:
You can add some `@brief` to `sample_idx` as well.

Reply:
Done.